ghcide/src/Development/IDE/Core/Compile.hs

```diff
@@ -901,9 +901,10 @@ indexHieFile se mod_summary srcPath !hash hf = do
             -- If the hash in the pending list doesn't match the current hash, then skip
             Just pendingHash -> pendingHash /= hash
         unless newerScheduled $ do
-          pre optProgressStyle
-          withHieDb (\db -> HieDb.addRefsFromLoaded db targetPath (HieDb.RealFile $ fromNormalizedFilePath srcPath) hash hf')
-          post
+          -- Using bracket, so even if an exception happen during withHieDb call,
+          -- the `post` (which clean the progress indicator) will still be called.
+          bracket_ (pre optProgressStyle) post $
+            withHieDb (\db -> HieDb.addRefsFromLoaded db targetPath (HieDb.RealFile $ fromNormalizedFilePath srcPath) hash hf')
```

Comment:
thinking about it more, I guess we should report to the user that an exception happened instead of silently ignoring it?

Reply:
Well, `bracket` does not ignore the exception. Right now it is propagated up, but before it ensures that `post` is called.

Actually, the behavior I had in my editor (neovim) was:

- Before: An exception is displayed in the editor, and then `Indexing` is stuck for an eternity
- Now: The exception is still displayed, but `Indexing` is not stuck anymore.

Reply:
OK, that makes sense. But it still seems wrong in `post` to just increment the counter and pretend nothing happened.

Perhaps we should also keep a record of the failures and report them separately when progress ends?

Ideally we would also retry on the failures, but that gets thorny very quickly. We don't want to keep retrying indefinitely! How have you observed this failing? Is it always just multiple HLS processes? We could add a lock to the `~/.cache/ghcide/...` directory to prevent starting multiple HLS processes on the same project.

I don't expect these to be resolved in this PR, but we should think more about all these interactions.

Reply:
```
HieDb writer thread SQLite error:\n  SQLite3 returned ErrorConstraint while attempting to perform step: UNIQUE constraint failed: defs.hieFile, defs.occ
```

Here is the exception which was returned up to my editor.

Reply:
I've opened https://github.com/haskell/haskell-language-server/issues/3622, thank you for remaining me to do it.